README.md
dcmt-cpp
========

[![CI](https://github.com/tueda/dcmt-cpp/workflows/CI/badge.svg?branch=master)](https://github.com/tueda/dcmt-cpp/actions?query=branch:master)
[![Documentation Status](https://readthedocs.org/projects/dcmt-cpp/badge/?version=latest)](https://dcmt-cpp.readthedocs.io/en/latest/?badge=latest)
[![Codecov](https://codecov.io/gh/tueda/dcmt-cpp/branch/master/graph/badge.svg)](https://codecov.io/gh/tueda/dcmt-cpp/branch/master)

A C++11 wrapper library for
[Dynamic Creator of Mersenne Twisters](https://github.com/MersenneTwister-Lab/dcmt).


Example
-------

```cpp
#include <iostream>
#include <dcmt/dcmt.h>

int main() {
  // Create 32-bit Mersenne Twister pseudo-random number generator
  // engines with a period 2^521-1. Their characteristic polynomials are
  // determined with id = 0, 1, 999 and generated sequences are highly
  // independent. The generator engines are initialized with seed =
  // 1234, 4567, 8901, respectively.
  dcmt::dcmt521 rng1{dcmt::dcmt521::param_type{0}, 1234};
  dcmt::dcmt521 rng2{dcmt::dcmt521::param_type{1}, 4567};
  dcmt::dcmt521 rng3{dcmt::dcmt521::param_type{999}, 8901};

  // Generate a number by using each generator engine.
  std::cout << rng1() << std::endl;
  std::cout << rng2() << std::endl;
  std::cout << rng3() << std::endl;

  return 0;
}
```


### CMake (3.14 or later)

```cmake
add_subdirectory(dcmt-cpp)
target_link_libraries(your_app PRIVATE dcmt)
```


Development
-----------

```bash
# Install prerequisites (minimal).
brew install cmake gcc git

# Install prerequisites (including optional ones).
brew install cmake doxygen gcc git lcov llvm pre-commit

# Install pre-commit hooks.
pre-commit install
pre-commit install --hook-type commit-msg

# Set CPM source cache.
export CPM_SOURCE_CACHE=$HOME/.cache/CPM  # for Linux
export CPM_SOURCE_CACHE=$HOME/Library/Caches/CPM  # for macOS

# Run linters.
pre-commit run --all-files

# Tests.
cmake -S . -B build/debug -DCMAKE_BUILD_TYPE=Debug
cmake --build build/debug --target check

# Benchmarking.
cmake -S . -B build/release -DCMAKE_BUILD_TYPE=Release -DBUILD_BENCHMARKING=ON
NANOBENCH_SUPPRESS_WARNINGS=1 cmake --build build/release --target bench

# Documents.
cmake -S . -B build/docs -DBUILD_TESTING=OFF
cmake --build build/docs --target doc

# Code coverage.
cmake -S . -B build/coverage -DCMAKE_BUILD_TYPE=Debug -DUSE_CODE_COVERAGE=ON
lcov -z -d build/coverage
cmake --build build/coverage --target check
lcov -c -d build/coverage -o build/coverage/coverage.info
lcov -r build/coverage/coverage.info '*/c++/*' -o build/coverage/coverage.info
lcov -r build/coverage/coverage.info '*/lib/*' -o build/coverage/coverage.info
lcov -r build/coverage/coverage.info '*/doctest/*' -o build/coverage/coverage.info
lcov -r build/coverage/coverage.info '*/tests/*' -o build/coverage/coverage.info
genhtml -o build/coverage/html build/coverage/coverage.info

# Compiler sanitizers.
cmake -S . -B build/sanitizer -DCMAKE_BUILD_TYPE=Debug -DUSE_SANITIZER=ON
cmake --build build/sanitizer --target check

# Clang-Tidy.
cmake -S . -B build/clang-tidy -DCMAKE_BUILD_TYPE=Debug -DUSE_CLANG_TIDY=ON
cmake --build build/clang-tidy --target check
```


License
-------

[MIT](https://github.com/tueda/dcmt-cpp/blob/master/LICENSE)


CHANGELOG.md
# Changelog

## [1.0.0](https://github.com/tueda/dcmt-cpp/releases/tag/v1.0.0) - 2020-10-02

### Added

- First release.


tests/test-expand.cpp
// Instantiate template classes; expand them as much as possible.

#include <dcmt/dcmt.h>

template class dcmt::dcmt_engine<std::uint_fast32_t, 32, 521>;

void foo(std::istream& is, std::ostream& os) {
  dcmt::dcmt521 a;
  dcmt::dcmt521 b;
  os << (a != b);
  os << a;
  is >> b;
}

int main() { return 0; }


tests/test-readme.cpp

#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <dcmt/dcmt.h>
#include <doctest/doctest.h>

#include <algorithm>
#include <array>

using dcmt::dcmt521;
using std::array;
using std::generate;

const int n = 100;

const array<dcmt521::result_type, n> expected_32_521_0_4172_1234{
    0x527415d6U, 0x77876c15U, 0xf09e6deaU, 0xd582955cU, 0x3bee0f14U,
    0xbbf15d5eU, 0xdb6d7f03U, 0x19fef1a1U, 0xd95d4addU, 0x17335683U,
    0x9ff34b23U, 0xcac40fd4U, 0xe8393380U, 0xecdefa75U, 0xcde293ecU,
    0xf8c26d8cU, 0x5782a03bU, 0x7a9f5637U, 0xafd8de1fU, 0x6e536955U,
    0x5d093870U, 0xc974ce02U, 0xfafcf1f1U, 0x4290453fU, 0x917343a5U,

    0x4b8854deU, 0xbb9afa59U, 0x221def7cU, 0xb1bd1c69U, 0xc5a0eefbU,
    0xd7e8450fU, 0x25137d34U, 0x48bf3bf9U, 0x97affa5aU, 0xa8e058b0U,
    0xfdf92601U, 0xc40e5d81U, 0xe6fe5936U, 0xe85f30c3U, 0x55854119U,
    0x652b1c86U, 0xcd359b65U, 0x696965f4U, 0x0e3ac05bU, 0x29c9cf0eU,
    0xe1aa8361U, 0xe32e7878U, 0x74304d00U, 0x06c54d22U, 0x00f21e88U,

    0xdc7fe873U, 0xf46bd84bU, 0xef77f649U, 0x39f90de6U, 0x1a120005U,
    0xf505a12eU, 0x8366f0d3U, 0xe83b7089U, 0xb4e2725bU, 0x37606f1eU,
    0xa715182aU, 0x6f9b11b7U, 0x3f0480a0U, 0xa45c8d70U, 0xcac23c36U,
    0xc094e795U, 0x031d2316U, 0x76df78dcU, 0x4d093faeU, 0xbaff149cU,
    0xe0229959U, 0x51a9777cU, 0x5ee8ddd8U, 0x7339234bU, 0x77c945b3U,

    0x3c1541b7U, 0xd2d09768U, 0x3095cd41U, 0xea424743U, 0x53cc83d1U,
    0x27cc6546U, 0xef085e98U, 0x229f8e6aU, 0x7810f6b0U, 0x3a221c4cU,
    0x87ffdf72U, 0xfbdc539bU, 0x368e61fbU, 0xe1518647U, 0x71ac535cU,
    0x31289d90U, 0xc44f9487U, 0x8660be22U, 0xaf540143U, 0xace75b1dU,
    0x52157678U, 0xe59fffc5U, 0xe71a1798U, 0x0337fadcU, 0xe1b5650dU,
};

const array<dcmt521::result_type, n> expected_32_521_1_4172_4567{
    0x00af2259U, 0xc7d8a3a7U, 0x7ec9c866U, 0x0f3c64baU, 0xabbcbd72U,
    0xad5b6275U, 0xc060150eU, 0x0e57a0a4U, 0x2d44a12fU, 0x62913251U,
    0x0e1e542bU, 0x01bc579bU, 0x0e775298U, 0x9327e768U, 0xd7f9dd9aU,
    0xc8722eabU, 0xbc24b101U, 0xf2e987acU, 0x54c2e82fU, 0x4eb1dffaU,
    0xfc90975eU, 0x6c38b66bU, 0x2eadb281U, 0x2ac99fbfU, 0x8dbc3656U,

    0xcfe0707bU, 0x4b1262b5U, 0xded27d51U, 0xdb6de64fU, 0x209da1cdU,
    0x3e677eceU, 0x73f3537cU, 0xb3eedaf2U, 0x6f048773U, 0x3fae9aa7U,
    0x51bbd234U, 0x78e6f450U, 0x4111d622U, 0xf8ca1b4dU, 0xc0a5fbb8U,
    0x7ab193f0U, 0x6a0291b9U, 0xaca957e0U, 0xcb13720bU, 0x787ce40fU,
    0x6fa399c5U, 0x616cf6d6U, 0x0d15f1faU, 0xb819fed3U, 0x0f047875U,

    0x67a19877U, 0x56527673U, 0xfb38c7c4U, 0xfc77ba8fU, 0x4ce2c3eaU,
    0x2a9809a3U, 0xed5f8746U, 0x8b78fb66U, 0x7b87d381U, 0xa4e9a082U,
    0xfea77146U, 0xe0f55c61U, 0x10743238U, 0xeb7757a1U, 0xa20053afU,
    0x33294901U, 0x677e28e8U, 0x18915a46U, 0xbb89ebdfU, 0x756ecd7dU,
    0x2c39878aU, 0x529d1b40U, 0x4cdd60b2U, 0xca2fd1c9U, 0xa6a03ad3U,

    0x6e69f2c0U, 0x87c6c615U, 0xa9489541U, 0xa8b41d59U, 0x8c456d2bU,
    0x691e9c77U, 0xa9982f46U, 0xa8a291b1U, 0xc1c6b53bU, 0xe8e8390fU,
    0x8644fad7U, 0xcf475280U, 0xd4fcd949U, 0xb70f2acaU, 0xa18a9f42U,
    0x6dd3a7a9U, 0x1c6b542eU, 0x8bdac4ffU, 0xebb6e211U, 0xfe4497dcU,
    0x53c846beU, 0x1280b77fU, 0x0d0c5adaU, 0x30fe5622U, 0x5a5e6c5eU,
};

const array<dcmt521::result_type, n> expected_32_521_999_4172_8901{
    0x25810b80U, 0xc608933fU, 0xa7277662U, 0x2315fe87U, 0x80e94e47U,
    0x5406d957U, 0xc4181b50U, 0x9b45ac86U, 0xdd75553bU, 0x1cc69638U,
    0x2df38e5eU, 0x0c02307aU, 0x1da2b946U, 0x7f01e8d0U, 0x520c478eU,
    0x8d477443U, 0x8f7e447fU, 0x1fd4bb0aU, 0x2c3d2393U, 0x8737b749U,
    0x7b32454fU, 0xc9a48bebU, 0x5970a1e0U, 0xf06bf3fbU, 0xdadc2511U,

    0x850b50a1U, 0x5d35c5a4U, 0x2f4ef8f2U, 0x32a20abcU, 0xbb7d4253U,
    0x04575ad6U, 0xc374ece8U, 0x6f61b7e9U, 0x0e005919U, 0xb80a04afU,
    0x566c2530U, 0xfde0815fU, 0xc97a5c8bU, 0x6571350eU, 0x787ef66cU,
    0x6534b52cU, 0x78575612U, 0x7bc7ebdbU, 0xb29aa7dbU, 0x233a226fU,
    0x2e819c46U, 0xda442de5U, 0xf913314cU, 0x43c91e5eU, 0x694579bbU,

    0x0673478dU, 0x75c652c9U, 0x7ef89a17U, 0x943d2184U, 0x0db5a625U,
    0x4af9ad34U, 0x94ec2cc4U, 0x3553b6bbU, 0xaf34d8b2U, 0x8c8143cfU,
    0x924dc2bcU, 0xe2260354U, 0x6370e06aU, 0x7e2f8fc0U, 0x53d283afU,
    0xd30b0e4fU, 0xf12fd06bU, 0x3d0945e1U, 0x88d29a4fU, 0xf6423f3dU,
    0x89c0c18eU, 0xe176492eU, 0x0e506860U, 0x27f9f812U, 0x5522adbdU,

    0x79be4578U, 0x487d706cU, 0x579a3f33U, 0x33882f93U, 0x0c01ca1cU,
    0xb6bc45b6U, 0xe4096877U, 0xc13e0b34U, 0x287ba3faU, 0xa803176aU,
    0xb25714f0U, 0xdacade1fU, 0xd010af0bU, 0x1078d5f8U, 0xa6237ee0U,
    0x757fa997U, 0xf570726bU, 0x0234f897U, 0xb6b280ecU, 0x7743484dU,
    0x2b3975edU, 0x7186d1a5U, 0x85c2a9ebU, 0x554a25c7U, 0x4f3242f6U,
};

TEST_CASE("sequences in readme") {
  dcmt521 rng1{dcmt521::param_type{0}, 1234};
  dcmt521 rng2{dcmt521::param_type{1}, 4567};
  dcmt521 rng3{dcmt521::param_type{999}, 8901};

  array<dcmt521::result_type, n> a1;
  array<dcmt521::result_type, n> a2;
  array<dcmt521::result_type, n> a3;

  generate(begin(a1), end(a1), rng1);
  generate(begin(a2), end(a2), rng2);
  generate(begin(a3), end(a3), rng3);

  for (int i = 0; i < n; i++) {
    CHECK_MESSAGE(a1[i] == expected_32_521_0_4172_1234[i], "i == " << i);
    CHECK_MESSAGE(a2[i] == expected_32_521_1_4172_4567[i], "i == " << i);
    CHECK_MESSAGE(a3[i] == expected_32_521_999_4172_8901[i], "i == " << i);
  }
}


tests/test-dcmt521.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <dcmt/dcmt.h>
#include <doctest/doctest.h>

#include <algorithm>
#include <array>
#include <iomanip>
#include <iostream>

using dcmt::dcmt521;
using std::array;
using std::generate;
using std::stringstream;

const int n = 100;

const array<dcmt521::result_type, n> expected_32_521_0_4172_3241{
    0x4def2d53U, 0x4284dd16U, 0x97e05113U, 0x3180fe45U, 0xef6dccc5U,
    0xfa19d824U, 0xcb3393d4U, 0xe88a7e32U, 0x8815ac11U, 0x7326765bU,
    0x8f639a75U, 0xdd2a0752U, 0xa97518b0U, 0x3cec73dfU, 0xed4ed842U,
    0xc30b947cU, 0x86faecbaU, 0x06787d11U, 0x495698f2U, 0x49e68a66U,
    0x93867996U, 0xe80d33bfU, 0xb91c64cfU, 0x2fde66eeU, 0x3fb734b9U,
    0x5904020cU, 0x967baf87U, 0x29490af9U, 0xc44dfadaU, 0x1422660dU,
    0xcd03cfd4U, 0x7bea482dU, 0x1c278092U, 0x33c2cba3U, 0x35455c0fU,
    0xf1bb44d8U, 0xd2436ec0U, 0xb363bd2fU, 0x7d68ff1bU, 0x0fd0c3acU,
    0xfd3fc8d2U, 0x67347ad0U, 0x3d8b8f7dU, 0x4399434dU, 0x6e9e50b5U,
    0x1e3bc81cU, 0x8abedd12U, 0x7528fe20U, 0xb3057114U, 0xeffb376aU,
    0x3c233abaU, 0x5bf7fe59U, 0xbba87469U, 0xa91599ecU, 0x9fe1380aU,
    0x89f14b0dU, 0x186e70ffU, 0x09d1b88fU, 0x11ec92a0U, 0x3320fe0fU,
    0xc44e20deU, 0x5bc7cb5bU, 0xc59a1f51U, 0xf0c09c97U, 0xda55d9aeU,
    0xe3e73cb7U, 0x9ca0fb0aU, 0xc8f538b4U, 0xdee6807fU, 0xdc14bdbcU,
    0xea3cdcdcU, 0x9545b37eU, 0xa1e4874cU, 0x042cededU, 0x05128c27U,
    0x01c91173U, 0x62f29c93U, 0x844f62e0U, 0xd5fb909cU, 0xe9eebc3bU,
    0xffc8c4e0U, 0xaf96a417U, 0xc977acaaU, 0x07fff398U, 0x6e50a176U,
    0x0b93f137U, 0x792a3ed0U, 0x9781a1b5U, 0xd55804ccU, 0x85af48eaU,
    0x595313b7U, 0x34321bdbU, 0x9ddb5d46U, 0xd66ee620U, 0x342f0829U,
    0xbda834d0U, 0x387ee323U, 0x3ddf69f4U, 0x06fc97e5U, 0x7a4862e2U,
};

TEST_CASE("properties") {
  CHECK_EQ(dcmt521::word_size, 32);
  CHECK_EQ(dcmt521::prime_exponent, 521);
  CHECK_EQ(dcmt521::min(), 0);
  CHECK_EQ(dcmt521::max(), 0xffffffffU);
}

TEST_CASE("default ctor") {
  dcmt521 rng{};
  array<dcmt521::result_type, n> a;
  generate(begin(a), end(a), rng);

  for (int i = 0; i < n; i++) {
    CHECK_MESSAGE(a[i] == expected_32_521_0_4172_3241[i], "i == " << i);
  }
}

TEST_CASE("ctor(lvalue-param)") {
  dcmt521::param_type param;

  {
    dcmt521 rng{param};
    array<dcmt521::result_type, n> a;
    generate(begin(a), end(a), rng);

    for (int i = 0; i < n; i++) {
      CHECK_MESSAGE(a[i] == expected_32_521_0_4172_3241[i], "i == " << i);
    }
  }

  // discard, comparison operators
  {
    dcmt521 rng1{param};
    dcmt521 rng2{param};

    CHECK_EQ(rng1, rng2);

    rng1.discard(17);  // n = p/w + 1 = 17

    CHECK_NE(rng1, rng2);

    rng1.discard(1);

    CHECK_NE(rng1, rng2);

    for (int i = 0; i < 18; i++) {
      rng2();
    }

    CHECK_EQ(rng1, rng2);
  }

  // serialization
  {
    stringstream buf;

    dcmt521 rng1{param};
    rng1.discard(10);
    buf << rng1;
    rng1.discard(10);

    dcmt521 rng2{param};
    buf >> rng2;
    rng2.discard(10);

    CHECK_EQ(rng1, rng2);
  }

  // self-assignment should be no problem
  {
    dcmt521 rng{param};
    rng = rng;

    array<dcmt521::result_type, n> a;
    generate(begin(a), end(a), rng);

    for (int i = 0; i < n; i++) {
      CHECK_MESSAGE(a[i] == expected_32_521_0_4172_3241[i], "i == " << i);
    }
  }

  // id must be <= 65535
  { CHECK_THROWS_AS(dcmt521::param_type{65536}, std::invalid_argument); }
}


benches/test-bench.cpp
#define ANKERL_NANOBENCH_IMPLEMENT
#include <dcmt/dcmt.h>
#include <nanobench.h>

#include <random>

using ankerl::nanobench::Bench;
using ankerl::nanobench::doNotOptimizeAway;
using dcmt::dcmt521;
using std::mt19937;
using std::mt19937_64;
using std::normal_distribution;
using std::uniform_real_distribution;

#define bench_init(bench, rng, name) \
  do {                               \
    bench.run(name, [&]() {          \
      auto r = rng;                  \
      doNotOptimizeAway(r);          \
    });                              \
  } while (false)

template <typename Rng>
void bench_gen_raw(Bench& bench, const Rng& rng, const char* name) {
  Rng r(rng);
  bench.run(name, [&]() {
    auto x = r();
    doNotOptimizeAway(x);
  });
}

template <typename Rng>
void bench_gen_double(Bench& bench, const Rng& rng, const char* name) {
  Rng r(rng);
  auto dist = uniform_real_distribution<>();
  bench.run(name, [&]() {
    auto x = dist(r);
    doNotOptimizeAway(x);
  });
}

template <typename Rng>
void bench_gen_gauss(Bench& bench, const Rng& rng, const char* name) {
  Rng r(rng);
  auto dist = normal_distribution<>();
  bench.run(name, [&]() {
    auto x = dist(r);
    doNotOptimizeAway(x);
  });
}

#define bench_set(b, bench_func)                    \
  do {                                              \
    bench_func(b, dcmt521(), "dcmt::dcmt521");      \
    bench_func(b, mt19937(), "std::mt19937");       \
    bench_func(b, mt19937_64(), "std::mt19937_64"); \
  } while (false)

int main() {
  Bench b;

  b.title("init");
  bench_set(b, bench_init);

  b.relative(true).warmup(1000).clockResolutionMultiple(100000);

  b.title("gen_raw");
  bench_set(b, bench_gen_raw);

  b.title("gen_double");
  bench_set(b, bench_gen_double);

  b.title("gen_gauss");
  bench_set(b, bench_gen_gauss);

  return 0;
}


benches/test-complexity.cpp
#define ANKERL_NANOBENCH_IMPLEMENT
#include <dcmt/dcmt.h>
#include <nanobench.h>

#include <random>

using ankerl::nanobench::Bench;
using ankerl::nanobench::doNotOptimizeAway;
using dcmt::dcmt_param;

template <int p>
void bench_init_complexity(Bench& bench) {
  int id = 0;
  bench.complexityN(p).run("dcmt::dcmt_param<p = N>", [&] {
    dcmt::dcmt_param<std::uint_fast32_t, 32, p> param(id);
    if (++id > 65535) {
      id = 0;
    }
    doNotOptimizeAway(id);
    doNotOptimizeAway(param);
  });
}

int main() {
  Bench b;
  b.title("asymptotic complexity");
  bench_init_complexity<521>(b);
  bench_init_complexity<607>(b);
  bench_init_complexity<1279>(b);
  bench_init_complexity<2203>(b);
  bench_init_complexity<2281>(b);
  bench_init_complexity<3217>(b);
  bench_init_complexity<4253>(b);
  bench_init_complexity<4423>(b);
  std::cout << b.complexityBigO() << std::endl;  // expected to be O(p^3)

  return 0;
}


include/dcmt/dcmt.h
#ifndef DCMT_CPP_DCMT_H
#define DCMT_CPP_DCMT_H

#include <cassert>
#include <cstddef>  // std::size_t
#include <cstdint>
#include <istream>
#include <limits>
#include <memory>  // std::unique_ptr
#include <ostream>
#include <type_traits>
#include <utility>  // std::move, std::swap

/**
 * Namespace for classes that wraps the DCMT library.
 */
namespace dcmt {

/**
 * Internal details.
 */
namespace detail {

struct mt_struct {
  std::uint32_t aaa;
  int mm, nn, rr, ww;
  std::uint32_t wmask, umask, lmask;
  int shift0, shift1, shiftB, shiftC;
  std::uint32_t maskB, maskC;
  int i;
  std::uint32_t* state;
};

void* alloc_memory(std::size_t size);
void free_memory(void* ptr);
void free_mt_struct(mt_struct* mts);

inline bool operator==(const mt_struct& a, const mt_struct& b) {
  if (a.aaa != b.aaa || a.mm != b.mm || a.nn != b.nn || a.rr != b.rr ||
      a.ww != b.ww || a.wmask != b.wmask || a.umask != b.umask ||
      a.lmask != b.lmask || a.shift0 != b.shift0 || a.shift1 != b.shift1 ||
      a.shiftB != b.shiftB || a.shiftC != b.shiftC || a.maskB != b.maskB ||
      a.maskC != b.maskC || a.i != b.i) {
    return false;
  }
  for (int i = 0; i < a.nn; i++) {
    if (a.state[i] != b.state[i]) {
      return false;
    }
  }
  return true;
}

template <class CharT, class Traits>
inline std::basic_ostream<CharT, Traits>& operator<<(
    std::basic_ostream<CharT, Traits>& os, const mt_struct& mt) {
  os << mt.aaa << ' ' << mt.mm << ' ' << mt.nn << ' ' << mt.rr << ' ' << mt.ww
     << ' ' << mt.wmask << ' ' << mt.umask << ' ' << mt.lmask << ' '
     << mt.shift0 << ' ' << mt.shift1 << ' ' << mt.shiftB << ' ' << mt.shiftC
     << ' ' << mt.maskB << ' ' << mt.maskC << ' ' << mt.i;
  for (int i = 0; i < mt.nn; i++) {
    os << ' ' << mt.state[i];
  }
  return os;
}

template <class CharT, class Traits>
inline std::basic_istream<CharT, Traits>& operator>>(
    std::basic_istream<CharT, Traits>& is, mt_struct& mt) {
  mt_struct tmp;  // NOLINT(cppcoreguidelines-pro-type-member-init)

  is >> tmp.aaa >> std::ws >> tmp.mm >> std::ws >> tmp.nn >> std::ws >>
      tmp.rr >> std::ws >> tmp.ww >> std::ws >> tmp.wmask >> std::ws >>
      tmp.umask >> std::ws >> tmp.lmask >> std::ws >> tmp.shift0 >> std::ws >>
      tmp.shift1 >> std::ws >> tmp.shiftB >> std::ws >> tmp.shiftC >> std::ws >>
      tmp.maskB >> std::ws >> tmp.maskC >> std::ws >> tmp.i;

  tmp.state =
      static_cast<std::uint32_t*>(alloc_memory(sizeof(std::uint32_t) * tmp.nn));

  for (int i = 0; i < mt.nn; i++) {
    is >> std::ws >> tmp.state[i];
  }

  std::swap(tmp, mt);

  free_memory(tmp.state);

  return is;
}

struct mt_struct_deleter {
  void operator()(mt_struct* ptr) { free_mt_struct(ptr); }
};

class dcmt_engine_impl {
 public:
  dcmt_engine_impl(int w, int p, int id, std::uint32_t seed);
  dcmt_engine_impl(const dcmt_engine_impl& impl);  // deep copy
  dcmt_engine_impl(dcmt_engine_impl&&) noexcept = default;
  ~dcmt_engine_impl() = default;
  dcmt_engine_impl& operator=(const dcmt_engine_impl& impl);  // deep copy
  dcmt_engine_impl& operator=(dcmt_engine_impl&&) noexcept = default;

  void sgenrand(std::uint32_t seed);
  std::uint32_t genrand();
  void discard(unsigned long long z);  // NOLINT(google-runtime-int)

  friend bool operator==(const dcmt_engine_impl& a, const dcmt_engine_impl& b) {
    // See https://bugs.llvm.org/show_bug.cgi?id=28480
    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
    assert(a.ptr_);
    // See https://bugs.llvm.org/show_bug.cgi?id=28480
    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
    assert(b.ptr_);
    return *a.ptr_ == *b.ptr_;
  }

  template <class CharT, class Traits>
  friend std::basic_ostream<CharT, Traits>& operator<<(
      std::basic_ostream<CharT, Traits>& os, const dcmt_engine_impl& impl) {
    // See https://bugs.llvm.org/show_bug.cgi?id=28480
    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
    assert(impl.ptr_);
    return os << *impl.ptr_;
  }

  template <class CharT, class Traits>
  friend std::basic_istream<CharT, Traits>& operator>>(
      std::basic_istream<CharT, Traits>& is, dcmt_engine_impl& impl) {
    // See https://bugs.llvm.org/show_bug.cgi?id=28480
    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
    assert(impl.ptr_);
    return is >> *impl.ptr_;
  }

 private:
  std::unique_ptr<mt_struct, mt_struct_deleter> ptr_;
};

}  // namespace detail

/**
 * Set of parameters for Mersenne Twister pseudo-random number generators.
 *
 * @tparam UIntType unsigned integral type for generators
 * @tparam WordSize word size for pseudo-random numbers
 * @tparam p Mersenne prime exponent
 */
template <class UIntType, std::size_t WordSize, std::size_t p>
class dcmt_param {
  static_assert(std::is_integral<UIntType>::value &&
                    std::is_unsigned<UIntType>::value,
                "UIntType must be an unsigned integral type");

  static_assert(WordSize == 31 || WordSize == 32, "WordSize must be 31 or 32");

  static_assert((WordSize == 31 &&
                 std::numeric_limits<UIntType>::max() >= 0x7fffffffU) ||
                    (WordSize == 32 &&
                     std::numeric_limits<UIntType>::max() >= 0xffffffffU),
                "size of UIntType must be equal to or larger than WordSize");

  static_assert(
      p == 521 || p == 607 || p == 1279 || p == 2203 || p == 2281 ||
          p == 3217 || p == 4253 || p == 4423 || p == 9689 || p == 9941 ||
          p == 11213 || p == 19937 || p == 21701 || p == 23209 || p == 44497,
      "p must be in the range of 521 to 44497 and 2^p-1 must be a prime");

  using impl = detail::dcmt_engine_impl;

 public:
  /**
   * Integral type for generators.
   */
  using result_type = UIntType;

  /**
   * Word size that determines the range of generated numbers.
   */
  static constexpr std::size_t word_size = WordSize;

  /**
   * Mersenne prime exponent.
   */
  static constexpr std::size_t prime_exponent = p;

  /**
   * Default id to search for the characteristic polynomial.
   */
  static constexpr result_type default_id = 0;

  /**
   * Default random seed to search for the characteristic polynomial.
   */
  static constexpr result_type default_seed = 4172;

  /**
   * Searches for a set of parameters for pseudo-random number generators.
   *
   * @param id to generate the characteristic polynomial. Must be <= 65535
   * @param seed random seed to generate the characteristic polynomial
   */
  // NOTE: currently, only lower 32 bits are used for the seed.
  explicit dcmt_param(result_type id = default_id,
                      result_type seed = default_seed)
      : impl_(word_size, prime_exponent, id, seed) {}

 private:
  impl impl_;

  template <class U, std::size_t W, std::size_t P>
  friend class dcmt_engine;
};

/**
 * Mersenne Twister pseudo-random number generator engine.
 *
 * @tparam UIntType unsigned integral type for the engine
 * @tparam WordSize word size for the generated numbers
 * @tparam p Mersenne prime exponent
 */
template <class UIntType, std::size_t WordSize, std::size_t p>
class dcmt_engine {
  using impl = detail::dcmt_engine_impl;

 public:
  /**
   * Parameter type for the engine.
   */
  using param_type = dcmt_param<UIntType, WordSize, p>;

  /**
   * Integral type generated by the engine.
   */
  using result_type = UIntType;

  /**
   * Word size that determines the range of numbers generated by the engine.
   */
  static constexpr std::size_t word_size = WordSize;

  /**
   * Mersenne prime exponent.
   */
  static constexpr std::size_t prime_exponent = p;

  /**
   * Default random seed.
   */
  static constexpr result_type default_seed = 3241;

  /**
   * Constructs the engine.
   *
   * @param seed random seed
   */
  explicit dcmt_engine(result_type seed = default_seed)
      : dcmt_engine(param_type(), seed) {}

  /**
   * Constructs the engine.
   *
   * @param param set of parameters
   * @param seed random seed
   */
  explicit dcmt_engine(const param_type& param, result_type seed = default_seed)
      : impl_(param.impl_) {
    this->seed(seed);
  }

  /**
   * Constructs the engine.
   *
   * @param param set of parameters
   * @param seed random seed
   */
  explicit dcmt_engine(param_type&& param, result_type seed = default_seed)
      : impl_(std::move(param.impl_)) {
    this->seed(seed);
  }

  /**
   * Reinitializes the engine by a new seed value.
   *
   * @param value random seed
   */
  // NOTE: currently, only lower 32 bits are used for the seed.
  void seed(result_type value = default_seed) { impl_.sgenrand(value); }

  /**
   * Returns the next pseudo-random number.
   *
   * @return generated value
   */
  result_type operator()() { return impl_.genrand(); }

  /**
   * Advances the internal state.
   *
   * @param z number of advances
   */
  // NOTE: intentionally use "unsigned long long" for consistency with the
  // standard <random> library.
  // NOLINTNEXTLINE(google-runtime-int)
  void discard(unsigned long long z) { impl_.discard(z); }

  /**
   * Returns the minimum value potentially generated by the engine, which is 0.
   *
   * @return minimum value
   */
  static constexpr result_type min() { return 0U; }

  /**
   * Returns the maximum value potentially generated by the engine.
   *
   * @return maximum value
   */
  static constexpr result_type max() { return 0xffffffffU >> (32 - word_size); }

  /**
   * Compares two engines.
   *
   * @param a first engine
   * @param b second engine
   * @return `true` if the engines are equivalent including their internal
   * states, `false` otherwise
   */
  friend bool operator==(const dcmt_engine& a, const dcmt_engine& b) {
    return a.impl_ == b.impl_;
  }

  /**
   * Compares two engines.
   *
   * @param a first engine
   * @param b second engine
   * @return `true` if the engines are not equivalent including their internal
   * states, `false` otherwise
   */
  friend bool operator!=(const dcmt_engine& a, const dcmt_engine& b) {
    return !(a == b);
  }

  /**
   * Serializes the state of the given engine into a stream.
   *
   * @param os output stream
   * @param e  engine to be serialized
   * @return `os`
   */
  template <class CharT, class Traits>
  friend std::basic_ostream<CharT, Traits>& operator<<(
      std::basic_ostream<CharT, Traits>& os, const dcmt_engine& e) {
    return os << e.impl_;
  }

  /**
   * Deserializes the state of the given engine from a stream.
   *
   * @param is input stream
   * @param e  engine to be deserialized
   * @return `is`
   */
  template <class CharT, class Traits>
  friend std::basic_istream<CharT, Traits>& operator>>(
      std::basic_istream<CharT, Traits>& is, dcmt_engine& e) {
    return is >> e.impl_;
  }

 private:
  impl impl_;
};

// Constant members.

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr std::size_t dcmt_param<UIntType, WordSize, p>::word_size;

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr std::size_t dcmt_param<UIntType, WordSize, p>::prime_exponent;

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr UIntType dcmt_param<UIntType, WordSize, p>::default_id;

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr UIntType dcmt_param<UIntType, WordSize, p>::default_seed;

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr std::size_t dcmt_engine<UIntType, WordSize, p>::word_size;

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr std::size_t dcmt_engine<UIntType, WordSize, p>::prime_exponent;

template <class UIntType, std::size_t WordSize, std::size_t p>
constexpr UIntType dcmt_engine<UIntType, WordSize, p>::default_seed;

// Predefined engine.

/**
 * 32-bit Mersenne Twister pseudo-random number generator engine with a period
 * 2^521-1.
 */
using dcmt521 = dcmt_engine<std::uint_fast32_t, 32, 521>;

}  // namespace dcmt

#endif  // DCMT_CPP_DCMT_H
